breadbox/breadbox/api/temp/associations.py

```diff
@@ -0,0 +1,120 @@
+from breadbox.config import Settings, get_settings
+from breadbox.db.session import SessionWithUser
+from breadbox.api.uploads import construct_file_from_ids
+from breadbox.schemas.associations import Associations
+from breadbox.schemas.associations import AssociationTable, AssociationsIn
+from typing import List
+from breadbox.service import associations as associations_service
+from breadbox.crud import associations as associations_crud
+import uuid
+from breadbox.db.util import transaction
+
+from .router import router
+
+
+@router.post(
+    "/associations/query-slice",
+    operation_id="get_associations_for_slice",
+    response_model=Associations,
+    response_model_exclude_none=False,
+)
+def query_associations_for_slice(
+    db: Annotated[SessionWithUser, Depends(get_db_with_user)],
+    settings: Annotated[Settings, Depends(get_settings)],
+    slice_query: Annotated[
+        SliceQuery, Body(description="A Data Explorer 2 context expression")
+    ],
+):
+    return associations_service.get_associations(
+        db, settings.filestore_location, slice_query
+    )
+
+
+@router.get(
+    "/associations",
+    operation_id="get_associations",
+    response_model=List[AssociationTable],
+)
+def get_associations(db: Annotated[SessionWithUser, Depends(get_db_with_user)]):
+    return [
+        AssociationTable(
+            id=a.id, dataset_1_id=a.dataset_1_id, dataset_2_id=a.dataset_2_id
+        )
+        for a in associations_crud.get_association_tables(db, None)
+    ]
+
+
+@router.delete("/associations/{id}")
+def delete_associations(
+    db: Annotated[SessionWithUser, Depends(get_db_with_user)],
+    settings: Annotated[Settings, Depends(get_settings)],
+    id: str,
+):
+    with transaction(db):
+        return associations_crud.delete_association_table(
+            db, id, settings.filestore_location
+        )
+
+
+@router.post(
+    "/associations",
+    operation_id="add_associations",
+    response_model=AssociationTable,
+    response_model_exclude_none=False,
+)
+def add_associations(
+    db: Annotated[SessionWithUser, Depends(get_db_with_user)],
+    settings: Annotated[Settings, Depends(get_settings)],
+    associations_in: Annotated[
+        AssociationsIn,
+        Body(
+            description="The associations table and which two datasets table references"
+        ),
+    ],
+):
+
+    serializer = URLSafeSerializer(settings.breadbox_secret)
+
+    full_file = construct_file_from_ids(
+        associations_in.file_ids,
+        associations_in.md5,
+        serializer,
+        settings.compute_results_location,
+    )
+
+    dest_filename = f"associations/{uuid.uuid4()}.sqlite3"
+    full_dest_filename = os.path.join(settings.filestore_location, dest_filename)
+    # make sure the directory exists before trying to move the file to it's final name
+    os.makedirs(os.path.dirname(full_dest_filename), exist_ok=True)
+
+    os.rename(full_file, full_dest_filename)
```

Comment:
I see that in one of the tests, you explicitly define the schema for the little sqlite file/database, but you don't do that here (in the add_associations endpoint where the file is created). Is that because the user is expected to upload a sqlite3 file with the correct schema?

I suppose that would make sense given that we already have these files created somewhere else and just need to copy them into breadbox, right?

Reply:
Yes, correct: my plan is to directly load the pre-existing sqlite3 files which are spit out from the preprocessing pipeline.

Reply:
Okay great!